### PR TITLE
Feature : First iteration of message queuing

### PR DESCRIPTION
**Pull request by ganfra**
## Type of change

- [x] Feature
- [ ] Bugfix
- [ ] Technical
- [ ] Other :

## Content

This is the first iteration of the message queuing, it includes : 
 - automatically retry failed requests as soon as the network is availble again
 - remove manual retry options, and fail sending state in the UI (this will be back soon..)
 - redacting a local echo will cancel sending it
 

## Motivation and context

Closes https://github.com/element-hq/element-x-android/issues/2863
Closes https://github.com/element-hq/element-x-android/issues/2862

## Tests

- Step 1
- Step 2
- Step ...

## Tested devices

- [ ] Physical
- [ ] Emulator
- OS version(s):

## Checklist

- [ ] Changes have been tested on an Android device or Android emulator with API 23
- [ ] UI change has been tested on both light and dark themes
- [ ] Accessibility has been taken into account. See https://github.com/element-hq/element-x-android/blob/develop/CONTRIBUTING.md#accessibility
- [ ] Pull request is based on the develop branch
- [ ] Pull request includes a new file under ./changelog.d. See https://github.com/element-hq/element-x-android/blob/develop/CONTRIBUTING.md#changelog
- [ ] Pull request includes screenshots or videos if containing UI changes
- [ ] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
- [ ] You've made a self review of your PR

### CODE REVIEW

File: libraries/matrix/api/src/main/kotlin/io/element/android/libraries/matrix/api/MatrixClient.kt [hunk 1]
```diff
@@ -105,4 +105,20 @@ interface MatrixClient : Closeable {
     suspend fun getRecentlyVisitedRooms(): Result<List<RoomId>>
     suspend fun resolveRoomAlias(roomAlias: RoomAlias): Result<ResolvedRoomAlias>
     suspend fun getRoomPreviewFromRoomId(roomId: RoomId, serverNames: List<String>): Result<RoomPreview>
+
+    /**
+     * Enables or disables the sending queue, according to the given parameter.
+     *
+     * The sending queue automatically disables itself whenever sending an
+     * event with it failed (e.g., sending an event via the Timeline,
```
Comment (jmartinesp) [inline libraries/matrix/api/src/main/kotlin/io/element/android/libraries/matrix/api/MatrixClient.kt:113]: Very tiny nit: unclosed parenthesis 😅 .

File: libraries/matrix/impl/src/main/kotlin/io/element/android/libraries/matrix/impl/timeline/RustTimeline.kt [hunk 2]
```diff
@@ -293,21 +313,26 @@ class RustTimeline(
...
+                when {
+                    originalEventId != null -> {
+                        val editedEvent = specialModeEventTimelineItem ?: inner.getEventTimelineItemByEventId(originalEventId.value)
+                        editedEvent.use {
+                            inner.edit(
+                                newContent = messageEventContentFromParts(body, htmlBody).withMentions(mentions.map()),
+                                editItem = it,
+                            )
+                        }
+                        specialModeEventTimelineItem = null
```
Comment (jmartinesp) [inline libraries/matrix/impl/src/main/kotlin/io/element/android/libraries/matrix/impl/timeline/RustTimeline.kt:326]: This isn't needed for the other when branch?

**Review by jmartinesp — APPROVED** [review]
LGTM after a quick review, just a couple of comments to double check.

File: libraries/matrix/impl/src/main/kotlin/io/element/android/libraries/matrix/impl/timeline/RustTimeline.kt (hunk 2 above)
Comment (ganfra) [reply libraries/matrix/impl/src/main/kotlin/io/element/android/libraries/matrix/impl/timeline/RustTimeline.kt:326]: Actually, we want to remove edition of local echo, as it'll break the ordering in the sending queue.
I'll update the api and usage.

**Pull request merged**